internal/api/option/error.go

```diff
@@ -0,0 +1,92 @@
+package option
+
+import (
+	"context"
+	"net/http"
+
+	"connectrpc.com/connect"
+	"github.com/pkg/errors"
+	kubeerr "k8s.io/apimachinery/pkg/api/errors"
+)
+
+var (
+	_ connect.Interceptor = &errorInterceptor{}
+)
+
+type errorInterceptor struct{}
+
+func newErrorInterceptor() connect.Interceptor {
+	return &errorInterceptor{}
+}
+
+func (i *errorInterceptor) WrapUnary(next connect.UnaryFunc) connect.UnaryFunc {
+	return func(ctx context.Context, req connect.AnyRequest) (connect.AnyResponse, error) {
+		res, err := next(ctx, req)
+		if err != nil {
+			return nil, i.unwrapError(err)
+		}
+		return res, nil
+	}
+}
+
+func (i *errorInterceptor) WrapStreamingClient(
+	next connect.StreamingClientFunc,
+) connect.StreamingClientFunc {
+	// TODO: Support streaming client when necessary
+	return next
+}
+
+func (i *errorInterceptor) WrapStreamingHandler(
+	next connect.StreamingHandlerFunc,
+) connect.StreamingHandlerFunc {
+	return func(ctx context.Context, conn connect.StreamingHandlerConn) error {
+		if err := next(ctx, conn); err != nil {
+			return i.unwrapError(err)
+		}
+		return nil
+	}
+}
+
+func (i *errorInterceptor) unwrapError(err error) error {
+	var connectErr *connect.Error
+	if ok := errors.As(err, &connectErr); !ok {
+		return err
+	}
+	underlyingErr := connectErr.Unwrap()
+	var statusErr *kubeerr.StatusError
+	if ok := errors.As(underlyingErr, &statusErr); ok {
+		return connect.NewError(httpStatusToConnectCode(statusErr.Status().Code), statusErr)
+	}
+	return err
+}
```

Comment:
@krancour, thank you for the review :)

After reading your review, I've realized that handling errors with interceptors might lead to confusion and make debugging difficult. For example, an interceptor could prioritize certain underlying errors, leaving the business layer (handler functions) with no control over the response code.

IMHO, it would be clearer if, instead of requiring all handler functions to return a `connect.Error` with a specific code and make interceptors to take care of panic-like errors only. WDYT?

https://github.com/akuity/kargo/pull/1266/commits/b823b71688147771f5c54c652c9de76849e86599 introduces `getCodeFromError`, allowing handlers to use it as a fallback (to extract status code) after attempting to handle errors with best effort.

Reply:
I liked the first attempt -- the filter based one -- better, to be honest.

`getCodeFromError` looks useful, but the problem with it is that one must always _remember_ to use it.

With the filter-based approach, the handler has control over the error code returned if it needs to it, and in cases where the handler chooses not to do that, the filter can make a good inference.

So the first approach was good, imho. Can we just cover the corner cases better?